### DISCUSSION TITLE
Apollo Cache missed when navigating between pages

### DISCUSSION THREAD

**Discussion by kennandavison** [vercel/next.js#12431] [Help]
Hi - I notice when I navigate using a link/router push between pages my Apollo cache is missed on my first request as if I've reloaded the page instead of navigating... Is this intended? Maybe this is a dev only thing? Haven't tried prod yet...

Ex: 2 pages with same query request when you hit a button, you hit it on page 1 (it makes network request and caches) and then navigate to page 2 and hit the button

Expected result: cache returns result
Actual result: network request on first request

**Comment by dhazeley** [comment]
I have been dealing with the same issue and have yet to find a solution. I'm using the new Apollo Client 3.0 and all my Next.js pages are completely static.

You would think that wrapping the custom `App` component in a HOC that holds the apollo client would persist data between client side routing, but it seems that my apollo client gets reinitialized on every page render.

**Reply by kennandavison** [reply]
@dhazeley Yeah, I'm thinking persisting the Apollo cache in localstorage might be the best solution for now... Not ideal though.

**Reply by antoniocapelo** [reply]
Hey, not sure if this helps, but recently I had this problem on a next project and the issue was that the `apolloClient` creation was being done at the `app` level. 

Using a `useRef` wrapping the creation and passing the `ref` to the `ApolloProvider` was the solution for this case.

**Reply by kennandavison** [reply]
@antoniocapelo - could you clarify "ref to the Apollo Provider?" Running into this same thing half a year later still haha.

**Reply by kennandavison** [reply]
https://github.com/apollographql/react-apollo/issues/3595 seems to be related, but the solution didn't work for me personally

**Reply by dhavall08** [reply]
I had a similar issue. It was resolved using `useMemo` and passing the server-side apollo `state` as a dependency, similar to this: [apollographql.com/blog/apollo-client/next-js/building-a-next-js-app-with-slash-graphql](apollographql.com/blog/apollo-client/next-js/building-a-next-js-app-with-slash-graphql/)
